binutils: update to 2.45

MingcongBai opened the pull request:
Topic Description
-----------------

- binutils: update to 2.45
    Co-authored-by: 白铭骢 \(Mingcong Bai\) \(@MingcongBai\) <jeffbai@aosc.io>
- gcc: Backport a change in gcc-15 development cycle to fix warnings with gas-2.45.
- gcc-13: Backport a change in gcc-15 development cycle to fix warnings with gas-2.45.
- gcc+cross: Backport a change in gcc-15 development cycle to fix warnings with gas-2.45.
    

Package(s) Affected
-------------------

- binutils: 2.45
- binutils+cross-amd64: 2.45
- binutils+cross-arm-none-eabi: 2.45
- binutils+cross-arm64: 2.45
- binutils+cross-loongarch64: 2.45
- binutils+cross-loongson3: 2.45
- binutils+cross-ppc64el: 2.45
- binutils+cross-riscv64: 2.45
- gcc: 14.3.0-1
- gcc-13: 13.4.0-1
- gcc+cross-amd64: 14.3.0+glibc2.40+3-1
- gcc+cross-arm64: 14.3.0+glibc2.40+3-1
- gcc+cross-loongarch64: 14.3.0+glibc2.40+3-1
- gcc+cross-loongson3: 14.3.0+glibc2.40+3-1
- gcc+cross-ppc64el: 14.3.0+glibc2.40+3-1
- gcc+cross-riscv64: 14.3.0+glibc2.40+3-1
- gcc+cross-arm-none-eabi: 14.3.0+glibc2.40+3-1

Security Update?
----------------

No

Build Order
-----------

```
#buildit binutils gcc gcc-13 gcc+cross
```

Test Build(s) Done
------------------

**Primary Architectures**

- [x] AMD64 `amd64`
- [x] AArch64 `arm64`
- [x] LoongArch 64-bit `loongarch64`

**Secondary Architectures**

- [x] Loongson 3 `loongson3`
- [x] PowerPC 64-bit (Little Endian) `ppc64el`
- [ ] RISC-V 64-bit `riscv64`

xry111 reviewed (changes requested):
Please hold this for a while, we just found a bug in the LoongArch linker: https://github.com/loongson-community/discussions/issues/99

xry111 commented:
It seems we also need https://sourceware.org/pipermail/binutils/2025-August/143177.html or our stripping process will break.

xry111 committed:
binutils: update to 2.45

- Drop the patch changing COMMONPAGESIZE for arm64 and ppc64.
  Per Alan Modra it's an old workaround for a RELRO bug where the code
  should use MAXPAGESIZE but incorrectly used COMMONPAGESIZE instead,
  and the bug is already fixed.
- Drop the patches already upstreamed for 2.45.
- Reword the patch adding -O2 to CPPFLAGS so it no longer seems making
  no sense.
- Retain the LoongArch symbol size fix which is upstreamed after 2.45.
- Track patches at AOSC-Tracking/binutils-gdb @ aosc/binutils-2_45
  (HEAD: 01c9e2a526453e97c2644cad2bd119f74957425a).

Co-authored-by: 白铭骢 (Mingcong Bai) (@MingcongBai) <jeffbai@aosc.io>
Co-authored-by: Xi Ruoyao <xry111@xry111.site>

xry111 commented:
<img width="1714" height="1418" alt="截图 2025-08-18 17-56-10" src="https://github.com/user-attachments/assets/a57c18eb-4fa5-401d-83bb-b0770fc34424" />

Smoke tested on LoongArch.

xry111 commented:
> It seems we also need https://sourceware.org/pipermail/binutils/2025-August/143177.html or our stripping process will break.

Not needed.

xry111 commented:
GCC-14 triggers a warning in the new GAS:

<img width="1714" height="1418" alt="截图 2025-08-18 18-27-30" src="https://github.com/user-attachments/assets/47bd9a1c-7fe9-4525-9979-3ac911bbc650" />

xry111 reviewed (changes requested):
We'll need to backport the GCC change to 13/14 in order to fix the issue.

xry111 committed:
gcc: Add a patch to fix warnings with gas-2.45

Tracking patches at AOSC-Tracking/gcc @ aosc/releases/gcc-14.3.0
(HEAD: f9317c51db8ffec9906367bb4e6af57226d825ce).

xry111 committed:
gcc-13: Add a patch to fix warnings with gas-2.45

Tracking patches at AOSC-Tracking/gcc @ aosc/releases/gcc-13.4.0
(HEAD: fd359c1febcaf6e3925b837a08853a88ecf50aee).

xry111 committed:
gcc+cross: Add a patch to fix warnings with gas-2.45

Tracking patches at AOSC-Tracking/gcc @ aosc/releases/gcc-14.3.0
(HEAD: f9317c51db8ffec9906367bb4e6af57226d825ce).

xry111 committed:
gcc+cross: Make sure the patches are applied on amd64

The patches in 01-amd64 are not applied for amd64 itself due to
FAIL_ARCH.

Note that in autobuild the serial number must start from 01 because of:

    if os.path.exists(dst_dir) and task.group_seq > 1:
        shutil.rmtree(dst_dir)

xry111 commented:
<img width="1656" height="1090" alt="截图 2025-08-24 23-38-46" src="https://github.com/user-attachments/assets/70759565-106e-45f3-947d-a890dde55da0" />
<img width="1446" height="594" alt="截图 2025-08-24 23-39-38" src="https://github.com/user-attachments/assets/93ceaa49-032f-42fc-901e-a07849bb2c10" />
